Source/Debugger/NSObject+TextureDebugger.mm

```diff
@@ -0,0 +1,293 @@
+//
+//  NSObject+TextureDebugger.mm
+//  Texture
+//
+//  Copyright (c) 2017-present, Pinterest, Inc.  All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <AsyncDisplayKit/ASAvailability.h>
+
+#if AS_TEXTURE_DEBUGGER
+
+#import <AsyncDisplayKit/AsyncDisplayKit.h>
+#import <AsyncDisplayKit/ASCollectionElement.h>
+#import <AsyncDisplaykit/ASRectTable.h>
+#import <AsyncDisplayKit/NSObject+TextureDebugger.h>
+#import <AsyncDisplayKit/TDDOMContext.h>
+
+#import <PonyDebugger/PDDOMTypes.h>
+
+#import <queue>
+
+// Constants defined in the DOM Level 2 Core: http://www.w3.org/TR/DOM-Level-2-Core/core.html#ID-1950641247
+static const int kPDDOMNodeTypeElement = 1;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface NSObject (TDDOMNodeGenerating)
+
++ (NSString *)td_nodeName;
+
+@end
+
+@implementation NSObject (TextureDebugger)
+
++ (NSString *)td_nodeName
+{
+  return @"object";
+}
+
+- (PDDOMNode *)td_generateDOMNodeWithContext:(TDDOMContext *)context
+{
+  NSNumber *nodeId = [context idForObject:self];
+  [context.idToFrameInWindow setRect:[self td_frameInWindow] forKey:nodeId];
+  
+  PDDOMNode *node = [[PDDOMNode alloc] init];
+  node.nodeType = @(kPDDOMNodeTypeElement);
+  node.nodeId = nodeId;
+  node.nodeName = [[self class] td_nodeName];
+  node.attributes = @[ @"description", self.debugDescription ];
+
+  NSMutableArray *nodeChildren = [NSMutableArray array];
+  for (id child in [self td_children]) {
+    [nodeChildren addObject:[child td_generateDOMNodeWithContext:context]];
+  }
+  node.children = nodeChildren;
+  node.childNodeCount = @(nodeChildren.count);
+  
+  return node;
+}
+
+- (CGRect)td_frameInWindow
+{
+  return CGRectNull;
+}
+
+- (NSArray *)td_children
+{
+  return @[];
+}
+
+@end
+
+@implementation UIApplication (TextureDebugger)
+
++ (NSString *)td_nodeName
+{
+  return @"application";
+}
+
+- (NSArray *)td_children
+{
+  return self.windows;
+}
+
+@end
+
+@implementation CALayer (TextureDebugger)
+
++ (NSString *)td_nodeName
+{
+  return @"layer";
+}
+
+- (PDDOMNode *)td_generateDOMNodeWithContext:(TDDOMContext *)context
+{
+  // For backing store of a display node (view/layer), let the node handle this job
+  ASDisplayNode *displayNode = ASLayerToDisplayNode(self);
+  if (displayNode) {
+    return [displayNode td_generateDOMNodeWithContext:context];
+  }
+  
+  return [super td_generateDOMNodeWithContext:context];
+}
+
+- (CGRect)td_frameInWindow
+{
+  return [self convertRect:self.bounds toLayer:nil];
```

Comment:
I could be wrong, but I don't think this is actually correct for layers, only views? There's nothing in the documentation that it coverts to window for CALayer. In fact what it says is super opaque:

```
If you specify nil for the l parameter, this method returns the original rect with an origin added to the layer's frame's origin.
```

I don't know what origin is… perhaps it's based on the window…